src/trs/rewrite.rs

```diff
@@ -304,11 +300,196 @@ impl TRS {
+    ///
+    /// let result = TRS::inverse_evaluation_helper(&r, &t, &mut rng);
+    ///
+    /// assert_eq!(result.display(), "A(C(x_))");
+    /// # }
+    /// ```
+    pub fn inverse_evaluation_helper<R: Rng>(rule: &Rule, t: &Term, rng: &mut R) -> Term {
+        let rhs_idx = rng.gen_range(0, rule.rhs.len());
+        TRS::replace_term_helper(t, &rule.rhs[rhs_idx], rule.lhs.clone())
+    }
+    /// Given two rules, attempts to apply one rule inversely to the other.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[macro_use] extern crate polytype;
+    /// # extern crate programinduction;
+    /// # extern crate rand;
+    /// # extern crate term_rewriting;
+    /// # use programinduction::trs::{TRS, Lexicon};
+    /// # use rand::{thread_rng};
+    /// # use term_rewriting::{Context, RuleContext, Signature, parse_rule, parse_term};
+    /// # fn main() {
+    /// let mut sig = Signature::default();
+    ///
+    /// let r = parse_rule(&mut sig, "A(B(x_ x_)) = D B(x_ x_)").expect("parse of A(B(x_ x_)) = D B(x_ x_)");
+    /// let rule = parse_rule(&mut sig, "C(y_) = B(y_ y_)").expect("parse of C(y_) = B(y_ y_)");
+    /// let mut rng = thread_rng();
+    ///
+    /// let result = TRS::inverse_evaluate_rule_helper(&rule, &r, &mut rng);
+    ///
+    /// assert_eq!(result.unwrap().pretty(), "A(C(x_)) = D C(x_)");
+    /// # }
+    /// ```
+    pub fn inverse_evaluate_rule_helper<R: Rng>(
+        rule: &Rule,
+        r: &Rule,
+        rng: &mut R,
+    ) -> Option<Rule> {
+        let lhs = TRS::inverse_evaluation_helper(rule, &r.lhs, rng);
+        let mut rhs: Vec<Term> = vec![];
+        for idx in 0..r.rhs.len() {
+            rhs.push(TRS::inverse_evaluation_helper(rule, &r.rhs[idx], rng));
+        }
+        Rule::new(lhs, rhs)
+    }
+    /// Selects two Rules from the TRS at random and atempts to inverse evaluate one rule on the other, if it
+    /// succeeds it takes that new rule and inserts it imediately after the background.
+    pub fn inverse_evaluate<R: Rng>(&self, rng: &mut R) -> Result<TRS, SampleError> {
+        let mut trs = self.clone();
+        let num_rules = self.len();
+        let num_background = self
+            .lex
+            .0
+            .read()
+            .expect("poisoned lexicon")
+            .background
+            .len();
+        if num_background >= num_rules - 1 {
+            return Ok(trs);
+        }
+        let ref_idx: usize = rng.gen_range(0, num_rules);
+        let mut target_idx: usize = rng.gen_range(num_background, num_rules);
+        while ref_idx == target_idx {
+            target_idx = rng.gen_range(num_background, num_rules);
+        }
+        let mut ref_rule = trs.utrs.rules[ref_idx].clone();
+        let rhs_idx = rng.gen_range(0, ref_rule.rhs.len());
+        let temp_rule = Rule::new(ref_rule.lhs, vec![ref_rule.rhs[rhs_idx].clone()]);
+        if temp_rule == None {
+            return Ok(trs);
+        }
+        ref_rule = temp_rule.unwrap();
```

Comment:
Basically everything to this point is boilerplate to try and get a single rule out of the TRS. Is there a helper function for sampling a single rule from a TRS? If not, perhaps it would be clearer to extract much of this code into a helper so that you could say something like:

``` rust
let ref_rule = trs.sample_rule()?;
if ref_rule.lhs.variables().len() >= ref_rule.rhs().variables().len() {
    ...
    }
}
```